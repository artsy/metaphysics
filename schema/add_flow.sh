#!/bin/bash
# Based on http://stackoverflow.com/questions/151677/tool-for-adding-license-headers-to-source-files

for i in */*.js
do
  if ! grep -q @flow $i
  then
    (echo "" & echo "/* @flow */") > flowificator
    cat flowificator $i >$i.new && mv $i.new $i
    rm flowificator
  fi
done