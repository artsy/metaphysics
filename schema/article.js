import { has } from 'lodash';
import positron from '../lib/loaders/positron';
import cached from './fields/cached';
import AuthorType from './author';
import Image from './image';
import date from './fields/date';
import { IDFields, NodeInterface } from './object_identification';
import {
  GraphQLString,
  GraphQLObjectType,
  GraphQLNonNull,
} from 'graphql';

const ArticleType = new GraphQLObjectType({
  name: 'Article',
  interfaces: [NodeInterface],
  isTypeOf: (obj) => has(obj, 'title') && has(obj, 'author'),
  fields: () => ({
    ...IDFields,
    cached,
    author: {
      type: AuthorType,
      resolve: ({ author }) => author,
    },
    href: {
      type: GraphQLString,
      resolve: ({ slug }) => `/article/${slug}`,
    },
    published_at: date,
    slug: {
      type: GraphQLString,
    },
    thumbnail_title: {
      type: GraphQLString,
    },
    thumbnail_teaser: {
      type: GraphQLString,
    },
    thumbnail_image: {
      type: Image.type,
      resolve: ({ thumbnail_image }) => Image.resolve(thumbnail_image),
    },
    title: {
      type: GraphQLString,
    },
    updated_at: date,
  }),
});

const Article = {
  type: ArticleType,
  description: 'An Article',
  args: {
    id: {
      type: new GraphQLNonNull(GraphQLString),
      description: 'The ID of the Article',
    },
  },
  resolve: (root, { id }) => positron(`articles/${id}`),
};

export default Article;
