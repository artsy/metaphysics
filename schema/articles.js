import positron from '../lib/loaders/positron';
import Article from './article';
import ArticleSorts from './sorts/article_sorts';
import {
  GraphQLString,
  GraphQLBoolean,
  GraphQLList,
} from 'graphql';

const Articles = {
  type: new GraphQLList(Article.type),
  description: 'A list of Articles',
  args: {
    published: {
      type: GraphQLBoolean,
      defaultValue: true,
    },
    show_id: {
      type: GraphQLString,
    },
    sort: ArticleSorts,
  },
  resolve: (root, options) => {
    return positron('articles', options).then(articles => articles.results);
  },
};

export default Articles;
