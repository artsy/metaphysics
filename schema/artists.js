import gravity from '../lib/loaders/gravity';
import Artist from './artist';
import ArtistSorts from './sorts/artist_sorts';
import {
  GraphQLList,
  GraphQLInt,
} from 'graphql';

const Artists = {
  type: new GraphQLList(Artist.type),
  description: 'A list of Artists',
  args: {
    page: {
      type: GraphQLInt,
      defaultValue: 1,
    },
    size: {
      type: GraphQLInt,
    },
    sort: ArtistSorts,
  },
  resolve: (root, options) => gravity('artists', options),
};

export default Artists;
