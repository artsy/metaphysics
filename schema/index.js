import Status from './status';
import Article from './article';
import Articles from './articles';
import Artwork from './artwork';
import Artworks from './artworks';
import Artist from './artist';
import Artists from './artists';
import ExternalPartner from './external_partner';
import Fair from './fair';
import Fairs from './fairs';
import Gene from './gene';
import HomePage from './home';
import OrderedSets from './ordered_sets';
import Profile from './profile';
import Partner from './partner';
import Partners from './partners';
import FilterPartners from './filter_partners';
import filterArtworks from './filter_artworks';
import PartnerCategory from './partner_category';
import PartnerCategories from './partner_categories';
import PartnerShow from './partner_show';
import PartnerShows from './partner_shows';
import Sale from './sale/index';
import Sales from './sales';
import SaleArtwork from './sale_artwork';
import Search from './search';
import Show from './show';
import TrendingArtists from './trending';
import Me from './me';
import CausalityJWT from './causality_jwt';
import Viewer from './viewer';
import ObjectIdentification from './object_identification';
import {
  GraphQLSchema,
  GraphQLObjectType,
} from 'graphql';

const schema = new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'RootQueryType',
    fields: {
      article: Article,
      articles: Articles,
      artwork: Artwork,
      artworks: Artworks,
      artist: Artist,
      artists: Artists,
      causality_jwt: CausalityJWT,
      external_partner: ExternalPartner,
      fair: Fair,
      fairs: Fairs,
      filter_partners: FilterPartners,
      filter_artworks: filterArtworks(),
      gene: Gene,
      home_page: HomePage,
      me: Me,
      node: ObjectIdentification.NodeField,
      ordered_sets: OrderedSets,
      partner: Partner,
      partner_categories: PartnerCategories,
      partner_category: PartnerCategory,
      partner_show: PartnerShow,
      partner_shows: PartnerShows,
      partners: Partners,
      profile: Profile,
      sale: Sale,
      sale_artwork: SaleArtwork,
      sales: Sales,
      search: Search,
      show: Show,
      status: Status,
      trending_artists: TrendingArtists,
      viewer: Viewer,
    },
  }),
});

export default schema;
