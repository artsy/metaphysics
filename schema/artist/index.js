// @ts-check
import type { GraphQLFieldConfig } from "graphql"
import { pageable, getPagingParameters } from "relay-cursor-paging"
import { connectionFromArraySlice, connectionDefinitions } from "graphql-relay"
import { assign, compact, defaults, first, has, reject, includes } from "lodash"
import { exclude } from "lib/helpers"
import cached from "schema/fields/cached"
import initials from "schema/fields/initials"
import { markdown, formatMarkdownValue } from "schema/fields/markdown"
import numeral from "schema/fields/numeral"
import Image from "schema/image"
import Article from "schema/article"
import Artwork, { artworkConnection } from "schema/artwork"
import PartnerArtist from "schema/partner_artist"
import Meta from "./meta"
import PartnerShow from "schema/partner_show"
import Show from "schema/show"
import Sale from "schema/sale/index"
import ArtworkSorts from "schema/sorts/artwork_sorts"
import ArticleSorts from "schema/sorts/article_sorts"
import PartnerShowSorts from "schema/sorts/partner_show_sorts"
import SaleSorts from "schema/sale/sorts"
import ArtistCarousel from "./carousel"
import ArtistStatuses from "./statuses"
import ArtistArtworksFilters from "./artwork_filters"
import positron from "lib/loaders/legacy/positron"
import total from "lib/loaders/legacy/total"
import { GravityIDFields, NodeInterface } from "schema/object_identification"
import { GraphQLObjectType, GraphQLBoolean, GraphQLString, GraphQLNonNull, GraphQLList, GraphQLInt } from "graphql"

const artistArtworkArrayLength = (artist, filter) => {
  let length
  if (first(filter) === "for_sale") {
    length = artist.forsale_artworks_count
  } else if (first(filter) === "not_for_sale") {
    length = artist.published_artworks_count - artist.forsale_artworks_count
  } else {
    length = artist.published_artworks_count
  }
  return length
}

// TODO: Fix upstream, for now we remove shows from certain Partner types
const blacklistedPartnerTypes = ["Private Dealer", "Demo", "Private Collector", "Auction"]
const showsWithBLacklistedPartnersRemoved = shows => {
  return reject(shows, show => {
    if (show.partner) {
      return includes(blacklistedPartnerTypes, show.partner.type)
    }
    if (show.galaxy_partner_id) {
      return false
    }
    return true
  })
}

// TODO Get rid of this when we remove the deprecated PartnerShow in favour of Show.
const ShowField = {
  args: {
    active: {
      type: GraphQLBoolean,
    },
    at_a_fair: {
      type: GraphQLBoolean,
    },
    is_reference: {
      type: GraphQLBoolean,
    },
    size: {
      type: GraphQLInt,
      description: "The number of PartnerShows to return",
    },
    solo_show: {
      type: GraphQLBoolean,
    },
    status: {
      type: GraphQLString,
    },
    top_tier: {
      type: GraphQLBoolean,
    },
    visible_to_public: {
      type: GraphQLBoolean,
    },
    sort: PartnerShowSorts,
  },
  resolve: ({ id }, options, request, { rootValue: { relatedShowsLoader } }) => {
    return relatedShowsLoader(
      defaults(options, {
        artist_id: id,
        sort: "-end_at",
      })).then(shows => showsWithBLacklistedPartnersRemoved(shows))
  },
}

export const ArtistType = new GraphQLObjectType({
  name: "Artist",
  interfaces: [NodeInterface],
  isTypeOf: obj => has(obj, "birthday") && has(obj, "artworks_count"),
  fields: () => {
    return {
      ...GravityIDFields,
      cached,
      alternate_names: {
        type: new GraphQLList(GraphQLString),
      },
      articles: {
        args: {
          sort: ArticleSorts,
          limit: {
            type: GraphQLInt,
          },
        },
        type: new GraphQLList(Article.type),
        resolve: ({ _id }, options, request, { rootValue: { articlesLoader } }) =>
          articlesLoader(
            defaults(options, {
              artist_id: _id,
              published: true,
            })
          ).then(({ results }) => results),
      },
      artists: {
        type: new GraphQLList(Artist.type), // eslint-disable-line no-use-before-define
        args: {
          size: {
            type: GraphQLInt,
            description: "The number of Artists to return",
          },
          exclude_artists_without_artworks: {
            type: GraphQLBoolean,
            defaultValue: true,
          },
        },
        resolve: ({ id }, options, request, { rootValue: { relatedMainArtistsLoader } }) =>
          relatedMainArtistsLoader(
            defaults(options, {
              artist: [id],
            })
          ),
      },
      artworks: {
        type: new GraphQLList(Artwork.type),
        args: {
          size: {
            type: GraphQLInt,
            description: "The number of Artworks to return",
          },
          page: {
            type: GraphQLInt,
          },
          sort: ArtworkSorts,
          published: {
            type: GraphQLBoolean,
            defaultValue: true,
          },
          filter: {
            type: new GraphQLList(ArtistArtworksFilters),
          },
          exclude: {
            type: new GraphQLList(GraphQLString),
          },
        },
        resolve: ({ id }, options, request, { rootValue: { artistArtworksLoader } }) =>
          artistArtworksLoader(id, options).then(exclude(options.exclude, "id")),
      },
      artworks_connection: {
        type: artworkConnection,
        args: pageable({
          sort: ArtworkSorts,
          filter: {
            type: new GraphQLList(ArtistArtworksFilters),
          },
          published: {
            type: GraphQLBoolean,
            defaultValue: true,
          },
        }),
        resolve: (artist, options, request, { rootValue: { artistArtworksLoader } }) => {
          // Convert `after` cursors to page params
          const { limit: size, offset } = getPagingParameters(options)
          // Construct an object of all the params gravity will listen to
          const { sort, filter, published } = options
          const gravityArgs = { size, offset, sort, filter, published }
          return artistArtworksLoader(artist.id, gravityArgs).then(artworks =>
            connectionFromArraySlice(artworks, options, {
              arrayLength: artistArtworkArrayLength(artist, filter),
              sliceStart: offset,
            })
          )
        },
      },
      bio: {
        type: GraphQLString,
        resolve: ({ nationality, years, hometown, location }) => {
          return compact([nationality, years, hometown, location ? `based in ${location}` : undefined]).join(", ")
        },
      },
      biography: {
        type: Article.type,
        description: "The Artist biography article written by Artsy",
        resolve: ({ _id }, options, request, { rootValue: { articlesLoader } }) =>
          articlesLoader({
            published: true,
            biography_for_artist_id: _id,
            limit: 1,
          }).then(articles => first(articles.results)),
      },
      biography_blurb: {
        args: assign(
          {
            partner_bio: {
              type: GraphQLBoolean,
              description: "If true, will return featured bio over Artsy one.",
              defaultValue: false,
            },
          },
          markdown().args
        ),
        type: new GraphQLObjectType({
          name: "ArtistBlurb",
          fields: {
            credit: {
              type: GraphQLString,
              resolve: ({ credit }) => credit,
            },
            text: {
              type: GraphQLString,
              resolve: ({ text }) => text,
            },
            partner_id: {
              type: GraphQLString,
              resolve: ({ partner_id }) => partner_id,
              description: "The partner id of the partner who submitted the featured bio.",
            },
          },
        }),
        resolve: ({ blurb, id }, { format, partner_bio }, request, { rootValue: { partnerArtistsLoader } }) => {
          if (!partner_bio && blurb && blurb.length) {
            return { text: formatMarkdownValue(blurb, format) }
          }
          return partnerArtistsLoader(id, {
            size: 1,
            featured: true,
          }).then(partner_artists => {
            if (partner_artists && partner_artists.length) {
              const { biography, partner } = first(partner_artists)
              return {
                text: formatMarkdownValue(biography, format),
                credit: `Submitted by ${partner.name}`,
                partner_id: partner.id,
              }
            }
            return { text: formatMarkdownValue(blurb, format) }
          })
        },
      },
      birthday: {
        type: GraphQLString,
      },
      blurb: {
        args: markdown().args,
        type: GraphQLString,
        resolve: ({ blurb }, { format }) => {
          if (blurb.length) {
            return formatMarkdownValue(blurb, format)
          }
        },
      },
      carousel: ArtistCarousel,
      contemporary: {
        type: new GraphQLList(Artist.type), // eslint-disable-line no-use-before-define
        args: {
          size: {
            type: GraphQLInt,
            description: "The number of Artists to return",
          },
          exclude_artists_without_artworks: {
            type: GraphQLBoolean,
            defaultValue: true,
          },
        },
        resolve: ({ id }, options, request, { rootValue: { relatedContemporaryArtistsLoader } }) =>
          relatedContemporaryArtistsLoader(
            defaults(options, {
              artist: [id],
            })
          ),
      },
      consignable: {
        type: GraphQLBoolean,
        deprecationReason: "Favor `is_`-prefixed boolean attributes",
      },
      counts: {
        type: new GraphQLObjectType({
          name: "ArtistCounts",
          fields: {
            artworks: numeral(({ published_artworks_count }) => published_artworks_count),
            follows: numeral(({ follow_count }) => follow_count),
            for_sale_artworks: numeral(({ forsale_artworks_count }) => forsale_artworks_count),
            partner_shows: numeral(({ partner_shows_count }) => partner_shows_count),
            related_artists: numeral(({ id }) => total(`related/layer/main/artists`, { artist: id })),
            articles: numeral(({ _id }) =>
              positron("articles", {
                artist_id: _id,
                published: true,
                limit: 1,
                count: true,
              }).then(({ count }) => count)
            ),
          },
        }),
        resolve: artist => artist,
      },
      deathday: {
        type: GraphQLString,
      },
      display_auction_link: {
        type: GraphQLBoolean,
        deprecationReason: "Favor `is_`-prefixed boolean attributes",
      },
      exhibition_highlights: {
        args: {
          size: {
            type: GraphQLInt,
            description: "The number of Shows to return",
            defaultValue: 5,
          },
        },
        type: new GraphQLList(Show.type),
        description: "Custom-sorted list of shows for an artist, in order of significance.",
        resolve: ({ id }, options, request, { rootValue: { relatedShowsLoader } }) => {
          return relatedShowsLoader({
            artist_id: id,
            sort: "-relevance,-start_at",
            is_reference: true,
            visible_to_public: false,
            size: options.size,
          }).then(shows => showsWithBLacklistedPartnersRemoved(shows))
        },
      },
      formatted_artworks_count: {
        type: GraphQLString,
        description: "A string showing the total number of works and those for sale",
        resolve: ({ published_artworks_count, forsale_artworks_count }) => {
          let totalWorks = null
          if (published_artworks_count) {
            totalWorks = published_artworks_count + (published_artworks_count > 1 ? " works" : " work")
          }
          const forSaleWorks = forsale_artworks_count ? forsale_artworks_count + " for sale" : null
          return forSaleWorks && totalWorks ? totalWorks + ", " + forSaleWorks : totalWorks
        },
      },
      formatted_nationality_and_birthday: {
        type: GraphQLString,
        description: `A string of the form "Nationality, Birthday (or Birthday-Deathday)"`,
        resolve: ({ birthday, nationality, deathday }) => {
          let formatted_bday = !isNaN(birthday) && birthday ? "b. " + birthday : birthday
          formatted_bday = formatted_bday && formatted_bday.replace(/born/i, "b.")

          if (!isNaN(deathday) && deathday && formatted_bday) {
            formatted_bday = `${formatted_bday.replace("b. ", "")}–${deathday.match(/\d+/)}`
          }
          if (nationality && formatted_bday) {
            return nationality + ", " + formatted_bday
          }
          return nationality || formatted_bday
        },
      },
      gender: {
        type: GraphQLString,
      },
      href: {
        type: GraphQLString,
        resolve: artist => `/artist/${artist.id}`,
      },
      has_metadata: {
        type: GraphQLBoolean,
        resolve: ({ blurb, nationality, years, hometown, location }) => {
          return !!(blurb || nationality || years || hometown || location)
        },
      },
      hometown: {
        type: GraphQLString,
      },
      image: Image,
      initials: initials("name"),
      is_consignable: {
        type: GraphQLBoolean,
        resolve: ({ consignable }) => consignable,
      },
      is_display_auction_link: {
        type: GraphQLBoolean,
        description: "Only specific Artists should show a link to auction results.",
        resolve: ({ display_auction_link }) => display_auction_link,
      },
      is_followed: {
        type: GraphQLBoolean,
        resolve: ({ id }, {}, request, { rootValue: { followedArtistLoader } }) => {
          if (!followedArtistLoader) return false
          return followedArtistLoader(id).then(({ is_followed }) => is_followed)
        },
      },
      is_public: {
        type: GraphQLBoolean,
        resolve: artist => artist.public,
      },
      is_shareable: {
        type: GraphQLBoolean,
        resolve: artist => artist.published_artworks_count > 0,
      },
      location: {
        type: GraphQLString,
      },
      meta: Meta,
      nationality: {
        type: GraphQLString,
      },
      name: {
        type: GraphQLString,
      },
      partner_artists: {
        type: new GraphQLList(PartnerArtist.type),
        args: {
          size: {
            type: GraphQLInt,
            description: "The number of PartnerArtists to return",
          },
        },
        resolve: ({ id }, options, request, { rootValue: { partnerArtistsLoader } }) =>
          partnerArtistsLoader(id, options),
      },
      partner_shows: {
        type: new GraphQLList(PartnerShow.type),
        deprecationReason: "Prefer to use shows attribute",
        ...ShowField,
      },
      public: {
        type: GraphQLBoolean,
        deprecationReason: "Favor `is_`-prefixed boolean attributes",
      },
      sales: {
        type: new GraphQLList(Sale.type),
        args: {
          live: {
            type: GraphQLBoolean,
          },
          is_auction: {
            type: GraphQLBoolean,
          },
          size: {
            type: GraphQLInt,
            description: "The number of Sales to return",
          },
          sort: SaleSorts,
        },
        resolve: ({ id }, options, request, { rootValue: { relatedSalesLoader } }) =>
          relatedSalesLoader(
            defaults(options, {
              artist_id: id,
              sort: "timely_at,name",
            })
          ),
      },
      shows: {
        type: new GraphQLList(Show.type),
        ...ShowField,
      },
      sortable_id: {
        type: GraphQLString,
        description: "Use this attribute to sort by when sorting a collection of Artists",
      },
      statuses: ArtistStatuses,
      years: {
        type: GraphQLString,
      },
    }
  },
})
const Artist: GraphQLFieldConfig<ArtistType, *> = {
  type: ArtistType,
  description: "An Artist",
  args: {
    id: {
      description: "The slug or ID of the Artist",
      type: new GraphQLNonNull(GraphQLString),
    },
  },
  resolve: (root, { id }, request, resolver) => {
    if (id.length === 0) {
      return null
    }
    const { artistLoader } = (resolver.rootValue: any)
    return artistLoader(id)
  },
}
export default Artist

export const artistConnection = connectionDefinitions({
  nodeType: Artist.type,
}).connectionType
