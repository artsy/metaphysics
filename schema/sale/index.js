import Artwork from "schema/artwork"
import Image from "schema/image/index"
import Profile from "schema/profile"
import SaleArtwork from "schema/sale_artwork"
import cached from "schema/fields/cached"
import date from "schema/fields/date"
import gravity from "lib/loaders/legacy/gravity"
import moment from "moment"
import { GravityIDFields } from "schema/object_identification"
import { amount } from "schema/fields/money"
import { exclude } from "lib/helpers"
import { map } from "lodash"

import {
  GraphQLString,
  GraphQLObjectType,
  GraphQLNonNull,
  GraphQLList,
  GraphQLBoolean,
  GraphQLInt,
  GraphQLFloat,
} from "graphql"

const BidIncrement = new GraphQLObjectType({
  name: "BidIncrement",
  fields: {
    amount: {
      type: GraphQLInt,
    },
    from: {
      type: GraphQLInt,
    },
    to: {
      type: GraphQLInt,
    },
  },
})

const BuyersPremium = new GraphQLObjectType({
  name: "BuyersPremium",
  fields: {
    ...GravityIDFields,
    amount: amount(({ cents }) => cents),
    cents: {
      type: GraphQLInt,
      resolve: ({ cents }) => cents,
    },
    percent: {
      type: GraphQLFloat,
    },
  },
})

const SaleType = new GraphQLObjectType({
  name: "Sale",
  fields: () => {
    return {
      ...GravityIDFields,
      cached,
      artworks: {
        type: new GraphQLList(Artwork.type),
        args: {
          page: {
            type: GraphQLInt,
            defaultValue: 1,
          },
          size: {
            type: GraphQLInt,
            defaultValue: 25,
          },
          all: {
            type: GraphQLBoolean,
            defaultValue: false,
          },
          exclude: {
            type: new GraphQLList(GraphQLString),
            description: "List of artwork IDs to exclude from the response (irrespective of size)",
          },
        },
        resolve: ({ id }, options) => {
          const invert = saleArtworks => map(saleArtworks, "artwork")

          if (options.all) {
            return gravity
              .all(`sale/${id}/sale_artworks`, options)
              .then(invert)
              .then(exclude(options.exclude, "id"))
          }

          return gravity(`sale/${id}/sale_artworks`, options)
            .then(invert)
            .then(exclude(options.exclude, "id"))
        },
      },
      associated_sale: {
        type: SaleType,
        resolve: ({ associated_sale }) => {
          if (associated_sale && associated_sale.id) {
            return gravity(`sale/${associated_sale.id}`)
          }
          return null
        },
      },
      auction_state: {
        type: GraphQLString,
        resolve: ({ auction_state }) => auction_state,
        deprecationReason: "Favor `status` for consistency with other models",
      },
      bid_increments: {
        type: new GraphQLList(BidIncrement),
        description: "A bid increment policy that explains minimum bids in ranges.",
        resolve: sale => gravity(`increments`, { key: sale.increment_strategy }).then(incs => incs[0].increments),
      },
      buyers_premium: {
        type: new GraphQLList(BuyersPremium),
        description: "Auction's buyer's premium policy.",
        resolve: sale => {
          if (!sale.buyers_premium) return null

          return map(sale.buyers_premium.schedule, item => ({
            cents: item.min_amount_cents,
            symbol: sale.currency,
            percent: item.percent,
          }))
        },
      },
      cover_image: {
        type: Image.type,
        resolve: ({ image_versions, image_url }) => Image.resolve({ image_versions, image_url }),
      },
      currency: {
        type: GraphQLString,
      },
      description: {
        type: GraphQLString,
      },
      eligible_sale_artworks_count: {
        type: GraphQLInt,
      },
      end_at: date,
      event_start_at: date,
      event_end_at: date,
      href: {
        type: GraphQLString,
        resolve: ({ id }) => `/auction/${id}`,
      },
      name: {
        type: GraphQLString,
      },
      is_auction: {
        type: GraphQLBoolean,
      },
      is_auction_promo: {
        type: GraphQLBoolean,
        resolve: ({ sale_type }) => sale_type === "auction promo",
      },
      is_closed: {
        type: GraphQLBoolean,
        resolve: ({ auction_state }) => auction_state === "closed",
      },
      is_open: {
        type: GraphQLBoolean,
        resolve: ({ auction_state }) => auction_state === "open",
      },
      is_live_open: {
        type: GraphQLBoolean,
        resolve: ({ auction_state, live_start_at }) => {
          const liveStart = moment(live_start_at)
          return auction_state === "open" && (moment().isAfter(liveStart) || moment().isSame(liveStart))
        },
      },
      is_preview: {
        type: GraphQLBoolean,
        resolve: ({ auction_state }) => auction_state === "preview",
      },
      is_registration_closed: {
        type: GraphQLBoolean,
        resolve: ({ registration_ends_at }) => moment().isAfter(registration_ends_at),
      },
      is_with_buyers_premium: {
        type: GraphQLBoolean,
        resolve: ({ buyers_premium }) => buyers_premium,
      },
      live_start_at: date,
      profile: {
        type: Profile.type,
        resolve: ({ profile }) => profile,
      },
      registration_ends_at: date,
      require_bidder_approval: {
        type: GraphQLBoolean,
      },
      sale_artworks: {
        type: new GraphQLList(SaleArtwork.type),
        args: {
          page: {
            type: GraphQLInt,
            defaultValue: 1,
          },
          size: {
            type: GraphQLInt,
            defaultValue: 25,
          },
          all: {
            type: GraphQLBoolean,
            defaultValue: false,
          },
        },
        resolve: ({ id }, options) => {
          if (options.all) {
            return gravity.all(`sale/${id}/sale_artworks`, options)
          }

          return gravity(`sale/${id}/sale_artworks`, options)
        },
      },
      sale_type: {
        type: GraphQLString,
      },
      start_at: date,
      status: {
        type: GraphQLString,
        resolve: ({ auction_state }) => auction_state,
      },
      sale_artwork: {
        type: SaleArtwork.type,
        args: {
          id: {
            type: new GraphQLNonNull(GraphQLString),
          },
        },
        resolve: (sale, { id }) => gravity(`sale/${sale.id}/sale_artwork/${id}`),
      },
      symbol: {
        type: GraphQLString,
      },
    }
  },
})

const Sale = {
  type: SaleType,
  description: "A Sale",
  args: {
    id: {
      type: new GraphQLNonNull(GraphQLString),
      description: "The slug or ID of the Sale",
    },
  },
  resolve: (root, { id }) => gravity(`sale/${id}`),
}

export default Sale
