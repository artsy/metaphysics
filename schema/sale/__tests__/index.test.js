import moment from "moment"
import schema from "schema"
import { fill } from "lodash"
import { runQuery, runAuthenticatedQuery } from "test/utils"

describe("Sale type", () => {
  let gravity
  const Sale = schema.__get__("Sale")

  const sale = {
    id: "foo-foo",
    _id: "123",
    currency: "$",
    is_auction: true,
    increment_strategy: "default",
  }

  beforeEach(() => {
    gravity = sinon
      .stub()
      .withArgs("sale/foo-foo")
      .returns(Promise.resolve(sale))
    Sale.__Rewire__("gravity", gravity)
  })

  afterEach(() => {
    Sale.__ResetDependency__("gravity")
  })

  describe("auction state", () => {
    const query = `
      {
        sale(id: "foo-foo") {
          _id
          is_preview
          is_open
          is_live_open
          is_closed
          is_registration_closed
          auction_state
          status
        }
      }
    `

    it("returns the correct values when the sale is closed", () => {
      sale.auction_state = "closed"
      return runQuery(query).then(data => {
        expect(data).toEqual({
          sale: {
            _id: "123",
            is_preview: false,
            is_open: false,
            is_live_open: false,
            is_closed: true,
            is_registration_closed: false,
            auction_state: "closed",
            status: "closed",
          },
        })
      })
    })

    it("returns the correct values when the sale is in preview mode", () => {
      sale.auction_state = "preview"
      return runQuery(query).then(data => {
        expect(data).toEqual({
          sale: {
            _id: "123",
            is_preview: true,
            is_open: false,
            is_live_open: false,
            is_closed: false,
            is_registration_closed: false,
            auction_state: "preview",
            status: "preview",
          },
        })
      })
    })

    it("returns the correct values when the sale is open", () => {
      sale.auction_state = "open"
      sale.live_start_at = moment().add(2, "days")
      return runQuery(query).then(data => {
        expect(data).toEqual({
          sale: {
            _id: "123",
            is_preview: false,
            is_open: true,
            is_live_open: false,
            is_closed: false,
            is_registration_closed: false,
            auction_state: "open",
            status: "open",
          },
        })
      })
    })

    it("returns the correct values when the sale is in live mode", () => {
      sale.auction_state = "open"
      sale.live_start_at = moment().subtract(2, "days")
      return runQuery(query).then(data => {
        expect(data).toEqual({
          sale: {
            _id: "123",
            is_preview: false,
            is_open: true,
            is_live_open: true,
            is_closed: false,
            is_registration_closed: false,
            auction_state: "open",
            status: "open",
          },
        })
      })
    })

    it("returns the correct values when sale registration is closed", () => {
      sale.auction_state = "open"
      sale.registration_ends_at = moment().subtract(2, "days")
      return runQuery(query).then(data => {
        expect(data).toEqual({
          sale: {
            _id: "123",
            is_preview: false,
            is_open: true,
            is_live_open: true,
            is_closed: false,
            is_registration_closed: true,
            auction_state: "open",
            status: "open",
          },
        })
      })
    })
  })

  describe("sale_artworks_connection", () => {
    it("returns data from gravity", () => {
      const query = `
        {
          sale(id: "foo-foo") {
            sale_artworks_connection(first: 10) {
              pageInfo {
                hasNextPage
              }
              edges {
                node {
                  id
                }
              }
            }
          }
        }
      `
      sale.eligible_sale_artworks_count = 20
      const rootValue = {
        saleArtworksLoader: () => Promise.resolve(fill(Array(sale.eligible_sale_artworks_count), { id: "some-id" })),
      }

      return runAuthenticatedQuery(query, rootValue).then(data => {
        expect(data).toMatchSnapshot()
      })
    })
  })

  describe("sale_artworks", () => {
    const SaleArtwork = schema.__get__("SaleArtwork")
    const saleArtworks = [
      {
        minimum_next_bid_cents: 400000,
        sale_id: "foo-foo",
      },
      {
        minimum_next_bid_cents: 20000,
        sale_id: "foo-foo",
      },
    ]

    beforeEach(() => {
      gravity = sinon.stub()
      gravity.withArgs("sale/foo-foo").returns(Promise.resolve(sale))
      gravity
        .withArgs("sale/foo-foo/sale_artworks", { page: 1, size: 25, all: false })
        .returns(Promise.resolve(saleArtworks))
      gravity.withArgs("increments", { key: "default" }).returns(
        Promise.resolve([
          {
            key: "default",
            increments: [
              {
                from: 0,
                to: 399999,
                amount: 5000,
              },
              {
                from: 400000,
                to: 1000000,
                amount: 10000,
              },
            ],
          },
        ])
      )
      Sale.__Rewire__("gravity", gravity)
      SaleArtwork.__Rewire__("gravity", gravity)
    })

    afterEach(() => {
      Sale.__ResetDependency__("gravity")
      SaleArtwork.__ResetDependency__("gravity")
    })

    it("returns data from gravity", () => {
      const query = `
        {
          sale(id: "foo-foo") {
            sale_artworks {
              bid_increments
            }
          }
        }
      `

      return runAuthenticatedQuery(query).then(data => {
        expect(data.sale.sale_artworks[0].bid_increments.slice(0, 5)).toEqual([400000, 410000, 420000, 430000, 440000])

        expect(data.sale.sale_artworks[1].bid_increments.slice(0, 5)).toEqual([20000, 25000, 30000, 35000, 40000])
      })
    })
  })

  describe("buyers premium", () => {
    it("returns a valid object even if the sale has no buyers premium", () => {
      const query = `
        {
          sale(id: "foo-foo") {
            _id
            buyers_premium {
              amount
              cents
            }
          }
        }
      `

      return runQuery(query).then(data => {
        expect(data).toEqual({
          sale: {
            _id: "123",
            buyers_premium: null,
          },
        })
      })
    })

    it("returns a valid object if there is a complete buyers premium", () => {
      sale.buyers_premium = {
        schedule: [
          {
            min_amount_cents: 10000,
            currency: "USD",
          },
        ],
      }

      const query = `
        {
          sale(id: "foo-foo") {
            _id
            buyers_premium {
              amount
              cents
            }
          }
        }
      `

      return runQuery(query).then(data => {
        expect(data).toEqual({
          sale: {
            _id: "123",
            buyers_premium: [
              {
                amount: "$100",
                cents: 10000,
              },
            ],
          },
        })
      })
    })
  })

  describe("associated sale", () => {
    const query = `
      {
        sale(id: "foo-foo") {
          _id
          associated_sale{
            id
          }
        }
      }
    `

    it("does not error, but returns null for associated sale", () => {
      return runQuery(query).then(data => {
        expect(data).toEqual({
          sale: {
            _id: "123",
            associated_sale: null,
          },
        })
      })
    })

    it("returns the associated sale", () => {
      sale.associated_sale = {
        id: "foo-foo",
      }
      return runQuery(query).then(data => {
        expect(data).toEqual({
          sale: {
            _id: "123",
            associated_sale: {
              id: "foo-foo",
            },
          },
        })
      })
    })
  })
})
