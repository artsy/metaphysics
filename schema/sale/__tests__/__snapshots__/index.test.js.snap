// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sale type sale_artworks_connection returns data from gravity 1`] = `
Object {
  "sale": Object {
    "sale_artworks_connection": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "some-id",
          },
        },
        Object {
          "node": Object {
            "id": "some-id",
          },
        },
        Object {
          "node": Object {
            "id": "some-id",
          },
        },
        Object {
          "node": Object {
            "id": "some-id",
          },
        },
        Object {
          "node": Object {
            "id": "some-id",
          },
        },
        Object {
          "node": Object {
            "id": "some-id",
          },
        },
        Object {
          "node": Object {
            "id": "some-id",
          },
        },
        Object {
          "node": Object {
            "id": "some-id",
          },
        },
        Object {
          "node": Object {
            "id": "some-id",
          },
        },
        Object {
          "node": Object {
            "id": "some-id",
          },
        },
      ],
      "pageInfo": Object {
        "hasNextPage": true,
      },
    },
  },
}
`;
